lightwood/mixers/helpers/plinear.py

```diff
@@ -0,0 +1,171 @@
+    Implementation of probabilistic weights via Linear function
+    '''
+    def __init__(self, in_features, out_features, bias=True):
+        """
+
+        :param in_features:  as name suggests
+        :param out_features: this essentially the number of neurons
+        :param bias: if you want a specific bias
+        """
+
+        super(PLinear ,self).__init__()
+
+        self.in_features = in_features
+        self.out_features = out_features
+
+        # these are the matrices that we will optimize for
+        self.sigma = Parameter(torch.Tensor(out_features, in_features))
+        self.mean = Parameter(torch.Tensor(out_features, in_features))
+
+        # there can be various ways to sample, given various distributions, we will stick with discrete normal as it is way faster
+        self.w_sampler = self.w_discrete_normal
+
+        if bias:
+            self.bias = Parameter(torch.Tensor(out_features))
+        else:
+            self.register_parameter('bias', None)
+
+        self.reset_parameters()
+
+        #make sure that we tell the graph that these two need to be optimized
+        self.sigma.requiresGrad = True # set requiresGrad to true!
+        self.mean.requiresGrad = True  # set requiresGrad to true!
+
+
+    def reset_parameters(self):
+        """
+        This sets the initial values for the distribution parameters, mean, sigma
+
+        """
+        if self.w_sampler == self.w_normal:
+            init.kaiming_uniform_(self.mean, a=math.sqrt(5)) # initial means (just as in original linear)
+            init.uniform_(self.sigma, a=0.01, b=0.1) # initial sigmas from 0.5-2
+        elif self.w_sampler == self.w_discrete_normal:
+            init.kaiming_uniform_(self.mean, a=math.sqrt(5))  # initial means (just as in original linear)
+            init.uniform_(self.sigma, a=0.05, b=0.2)
+
+        if self.bias is not None:
+            fan_in, _ = init._calculate_fan_in_and_fan_out(self.mean)
+            bound = 1 / math.sqrt(fan_in)
+            init.uniform_(self.bias, -bound, bound)
+
+    def w_discrete_normal(self):
+        """
+        Sample a w matrix based on a discrete normal distribution
+        :return: w
+        """
+
+        bucket_seed = random.uniform(0,1) # gett a number that can help determines which bucket of the distribution can numbers fall in
+        sigma_multiplier = 1 # this tell it how wide do we stretch the random saples
+
+        # set the widths for sampling, based on the seed
+        if bucket_seed > 0.33:
+            sigma_multiplier = 2
+            if bucket_seed > 0.8:
+                sigma_multiplier = 4
+
+        # generate the initial tensor, this will ultimately transforms in to the weights
+        w = torch.Tensor(self.out_features, self.in_features)
+        # make sure that they are evently distributed between -1, 1
+        init.uniform_(w, a=-1, b=1)
+
+        # adjust based on sigma
+        w = self.mean*( 1+ w * torch.abs(self.sigma) * sigma_multiplier )
```

Comment:
How do we go about extracting some sort of confidence / interval of possible prediction in the last layer (which is presumably a big part of the goal here). Do we add an option to remove the `self.mean*` operation for the last layer and then look at the probability distribution here ?

Reply:
you need to sample results at the end to calculate the probability distribution of the net outcome